migrations: persist progress during intents migration

dt opened the pull request:
Release note: none.
Release justification: improve new functionality.

dt committed:
migrations: persist progress during intents migration

Release note: none.
Release justification: improve new functionality.

dt committed:
migrations: update displayed progress percent in separated intents

Release note: none.

Release justification: new functionality improvement.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/69679)
<!-- Reviewable:end -->

dt commented:
first commit is #69670, and rest are spit up for commit-by-commit review

dt commented:
okay, %-complete updates back in a way that gets along with mocking this time (though unfortunately now delays starting the real work while it counts, rather than doing it in parallel). A follow-up could pass two range iterators to allow fixing that but I don't know if I'll get to that right away.

itsbilal commented on pkg/migration/migrations/separated_intents.go:
```diff
@@ -301,17 +308,35 @@ func (m *migrateLockTablePool) run(ctx context.Context, workerIdx int) {
+				m.mu.Lock()
+				m.mu.lowWater[workerIdx] = retryRequest.rangeStart
+				m.mu.Unlock()
 			}
```
You can do a little better in the case of a nextReq with a resume span (a rare but still possible case):

```suggestion
} else if nextReq.barrierDone {
 // could use a better bool name to denote that this was a resumeSpan
 m.mu.Lock()
 m.mu.lowWater[workerIdx] = retryRequest.start
 m.mu.Unlock()
}
```

itsbilal commented on pkg/migration/migrations/separated_intents.go:
```diff
@@ -66,6 +68,7 @@ const migrateLockTableRetries = 3
 type migrateLockTableRequest struct {
 	start, end   roachpb.Key
 	debugRangeID roachpb.RangeID
+	rangeStart   roachpb.RKey // for progress.
```
I don't see this set anywhere, also why not reuse `start`, as that's the one we modify when we get a `resumeSpan` too?

dt replied:
start here was the actual key we were sending the requests to -- i.e. sometimes the local, sometimes the global. rangeStart here was the key for the range, i.e. what we'd seek to.

itsbilal commented on pkg/migration/migrations/separated_intents.go:
```diff
@@ -360,6 +403,18 @@ func (m *migrateLockTablePool) runStatusLogger(ctx context.Context) {
+func countTotalRanges(ctx context.Context, ri rangeIterator) (uint64, error) {
+	var total uint64
+	rs := roachpb.RSpan{Key: roachpb.RKeyMin, EndKey: roachpb.RKeyMax}
+	for ri.Seek(ctx, roachpb.RKeyMin, kvcoord.Ascending); ri.Valid(); ri.Next(ctx) {
```
You could replace this with `distSender.CountRanges()`.

dt replied:
Heh, I _could_ but I don't have a sender capable of counting in the tests, where we mock it.

itsbilal commented on pkg/migration/migrations/separated_intents.go:
```diff
@@ -347,10 +373,27 @@ func (m *migrateLockTablePool) runStatusLogger(ctx context.Context) {
+					prog := details.(*jobspb.Progress_Migration).Migration
+					prog.Watermark = lowWaterMark
+					if total > 0 && finished <= total {
+						return float32(finished) / float32(total)
```
Btw we spin off 2 requests per range, so `finished` will actually end at `2*total`. Might have to update this fraction as well as the log line.

dt replied:
this made me realize that I actually needed to change it up a little so we only consider a range "done" if we do both. Added a new commit that switches the channel to have ranges on it as the unit of work, then the workers do both the local and global to "complete" that unit, rather than doing them separately. That makes it easier to track what is done and what isn't, and makes the number just num ranges now, not something between numRanges and numRanges*2 based on which do/don't have global keys.

itsbilal commented on pkg/migration/migrations/separated_intents.go:
```diff
@@ -347,10 +373,27 @@ func (m *migrateLockTablePool) runStatusLogger(ctx context.Context) {
-			log.Infof(ctx, "%d ranges have completed lock table migration", finished)
+			total := m.total // TODO(dt): count in background and load via atomic.
+			lowWaterMark := m.lowWaterMark()
+			log.Infof(ctx, "%d of %d ranges, up to %s, have completed lock table migration", finished, total, lowWaterMark)
```
2x factor (see below)

itsbilal commented on pkg/migration/migrations/separated_intents.go:
```diff
@@ -407,7 +482,7 @@ func runSeparatedIntentsMigration(
 
 func separatedIntentsMigration(
-	ctx context.Context, cv clusterversion.ClusterVersion, deps migration.SystemDeps, _ *jobs.Job,
+	ctx context.Context, cv clusterversion.ClusterVersion, deps migration.SystemDeps, job *jobs.Job,
```
Can we add `job` to `migration.SystemDeps`?

dt replied:
It didn't feel like it belonged there to me: mentioned on #69670 that the `SystemDeps` is a static bag of dependency APIs/services that is passed to any migration, while the job is an argument to the individual migration, just like `cv`

itsbilal reviewed (commented):
Looking really solid!

dt committed:
migrations: put whole ranges on intent migration ch

This put ranges on the channel, instead of one or two spans derived from those ranges,
for the local and potentially global spans of those ranges.

This means that when a unit of work pulled from the channel is completed by a worker,
completing that unit means completing both spans, and when it is done, that range can
be considered done, whereas before, each unit of work -- the local and the global --
would need to have been completed for that range to be done. Tracking that would be much
more complicated.

Release note: none.
Release justification: none.

itsbilal commented on pkg/migration/migrations/separated_intents.go:
```diff
@@ -90,17 +98,21 @@ type migrateLockTablePool struct {
 	done     chan bool
 	status   [concurrentMigrateLockTableRequests]int64
 	finished uint64
```
Oversight on my part, but we could probably benefit with a comment on `finished` saying it should be updated atomically.

itsbilal reviewed (approved):
Looks great to me! Thanks for doing this. Just one nit that you can ignore if you want.

dt committed:
migrations: rename finished to finishedAtomic

Release note: none.
Release justification: none.

dt commented:
TFTRs!

bors r+
